src/realm/object-store/impl/deep_change_checker.cpp

```diff
@@ -0,0 +1,345 @@
+        }
+    }
+    find_all_related_tables(out, table, tables_in_filters);
+}
+
+DeepChangeChecker::DeepChangeChecker(TransactionChangeInfo const& info, Table const& root_table,
+                                     std::vector<RelatedTable> const& related_tables,
+                                     std::vector<KeyPathArray> key_path_arrays)
+    : m_info(info)
+    , m_root_table(root_table)
+    , m_key_path_arrays(key_path_arrays)
+    , m_root_object_changes([&] {
+        auto it = info.tables.find(root_table.get_key().value);
+        return it != info.tables.end() ? &it->second : nullptr;
+    }())
+    , m_related_tables(related_tables)
+{
+    // If all callbacks do have a filter, every `KeyPathArray` will have entries.
+    // In this case we need to check the `ColKey`s and pass the filtered columns
+    // to the checker.
+    // If at least one `NotificationCallback` does not have a filter we notify on any change.
+    // This is signaled by leaving the `m_filtered_columns_in_root_table` and
+    // `m_filtered_columns` empty.
+    auto all_callbacks_filtered = all_of(begin(m_key_path_arrays), end(m_key_path_arrays), [](auto key_path_array) {
+        return key_path_array.size() > 0;
+    });
+    if (all_callbacks_filtered) {
+        for (auto key_path_array : m_key_path_arrays) {
+            for (auto key_path : key_path_array) {
+                if (key_path.size() != 0) {
+                    m_filtered_columns_in_root_table.push_back(key_path[0].second);
+                }
+                for (auto key_path_element : key_path) {
+                    m_filtered_columns.push_back(key_path_element.second);
+                }
+            }
+        }
+    }
+}
+
+bool DeepChangeChecker::check_outgoing_links(Table const& table, int64_t object_key,
+                                             std::vector<ColKey> filtered_columns, size_t depth)
+{
+    auto table_key = table.get_key();
+
+    // First we create an iterator pointing at the table identified by `table_key` within the `m_related_tables`.
+    auto it = find_if(begin(m_related_tables), end(m_related_tables), [&](auto&& tbl) {
+        return tbl.table_key == table_key;
+    });
+    // If no iterator could be found the table is not contained in `m_related_tables` and we cannot check any
+    // outgoing links.
+    if (it == m_related_tables.end())
+        return false;
+    // Likewise if the table could be found but does not have any (outgoing) links.
+    if (it->links.empty())
+        return false;
+
+    // Check if we're already checking if the destination of the link is
+    // modified, and if not add it to the stack
+    auto already_checking = [&](int64_t col) {
+        auto end = m_current_path.begin() + depth;
+        auto match = std::find_if(m_current_path.begin(), end, [&](auto& p) {
+            return p.object_key == object_key && p.col_key == col;
+        });
+        if (match != end) {
+            for (; match < end; ++match)
```

Comment:
style nit: enclose the body of a for loop in curly-braces even if it's just one line.